packages/ember-metal/lib/meta_listeners.js

```diff
@@ -0,0 +1,154 @@
+/*
+ When we render a rich template hierarchy, the set of events that
+ *might* happen tends to be much larger than the set of events that
+ actually happen. This implies that we should make listener creation &
+ destruction cheap, even at the cost of making event dispatch more
+ expensive.
+
+ Thus we store a new listener with a single push and no new
+ allocations, without even bothering to do deduplication -- we can
+ save that for dispatch time, if an event actually happens.
+ */
+
+/* listener flags */
+export var ONCE = 1;
+export var SUSPENDED = 2;
+
+export var protoMethods = {
+
+  addToListeners(eventName, target, method, flags) {
+    if (!this._listeners) {
+      this._listeners = [];
+    }
+    this._listeners.push(eventName, target, method, flags);
+  },
+
+  _finalizeListeners() {
+    if (this._listenersFinalized) { return; }
+    if (!this._listeners) { this._listeners = []; }
+    let pointer = this.parent;
+    while (pointer) {
+      let listeners = pointer._listeners;
+      if (listeners) {
+        this._listeners = this._listeners.concat(listeners);
+      }
+      if (pointer._listenersFinalized) { break; }
+      pointer = pointer.parent;
+    }
+    this._listenersFinalized = true;
+  },
+
+  removeFromListeners(eventName, target, method, didRemove) {
+    let pointer = this;
+    while (pointer) {
+      let listeners = pointer._listeners;
+      if (listeners) {
+        for (let index = listeners.length - 4; index >= 0; index -= 4) {
+          if (listeners[index] === eventName && (!method || (listeners[index + 1] === target && listeners[index + 2] === method))) {
+            if (pointer === this) {
+              // we are modifying our own list, so we edit directly
+              if (typeof didRemove === 'function') {
+                didRemove(eventName, target, listeners[index + 2]);
+              }
+              listeners.splice(index, 4);
```

Comment:
Yes, it would make the code way less crazy. I would guess the array pattern (which I just kept from the previous version) was an attempt to avoid object allocations? Is there any evidence that actually helps?